File: apps/web/src/types/tasks.ts
import type { AgentRun, Artifact, AssigneeType, DevTaskStatus } from './shared'

export type EpicStatus = 'proposed' | 'ready' | 'in_progress' | 'blocked' | 'completed'
export type EpicPriority = 'low' | 'medium' | 'high'

export interface EpicCreate {
  title: string
  requirement_id?: string | null
  description?: string
  priority?: EpicPriority
  sequence_order?: number
  acceptance_criteria?: string[]
  business_goal?: string
  assignee_type?: AssigneeType
  assignee_id?: string | null
  assignee_name?: string | null
}

export interface EpicUpdate {
  title?: string
  description?: string
  status?: EpicStatus
  priority?: EpicPriority
  sequence_order?: number
  acceptance_criteria?: string[]
  business_goal?: string
  assignee_type?: AssigneeType
  assignee_id?: string | null
  assignee_name?: string | null
}

export interface Epic {
  id: string
  project_id: string
  requirement_id: string | null
  title: string
  description: string
  status: EpicStatus
  priority: EpicPriority
  sequence_order: number
  acceptance_criteria: string[]
  business_goal: string
  assignee_type: AssigneeType
  assignee_id: string | null
  assignee_name: string | null
  created_at: string
  updated_at: string
}

export type DevTaskType =
  | 'backend' | 'frontend' | 'full_stack' | 'testing'
  | 'documentation' | 'infrastructure' | 'refactor' | 'unknown'
export type DevTaskPriority = 'low' | 'medium' | 'high'

export interface DevTask {
  id: string
  project_id: string
  requirement_id: string | null
  ticket_id: string | null
  source_analysis_id: string | null
  agent_run_id: string
  epic_id: string | null
  title: string
  description: string
  task_type: DevTaskType
  status: DevTaskStatus
  priority: DevTaskPriority
  sequence_order: number
  depends_on: string[]
  acceptance_criteria: string[]
  definition_of_done: string[]
  qa_required: boolean
  suggested_agent_type: string | null
  assignee_type: AssigneeType
  assignee_id: string | null
  assignee_name: string | null
  created_at: string
  updated_at: string
  is_ready?: boolean
  blocked_by?: string[]
}

export interface DevTaskUpdate {
  title?: string
  description?: string
  status?: DevTaskStatus
  priority?: DevTaskPriority
  sequence_order?: number
  depends_on?: string[]
  acceptance_criteria?: string[]
  definition_of_done?: string[]
  qa_required?: boolean
  suggested_agent_type?: string | null
  epic_id?: string | null
  assignee_type?: AssigneeType
  assignee_id?: string | null
  assignee_name?: string | null
}

export interface SubtaskUpdate {
  title?: string
  description?: string
  status?: DevTaskStatus
  sequence_order?: number
  acceptance_criteria?: string[]
  qa_required?: boolean
  assignee_type?: AssigneeType
  assignee_id?: string | null
  assignee_name?: string | null
}

export interface Subtask {
  id: string
  dev_task_id: string
  project_id: string
  title: string
  description: string
  status: DevTaskStatus
  sequence_order: number
  acceptance_criteria: string[]
  qa_required: boolean
  assignee_type: AssigneeType
  assignee_id: string | null
  assignee_name: string | null
  created_at: string
  updated_at: string
}

export interface TaskDecompositionResponse {
  agent_run: AgentRun
  artifact: Artifact
  dev_tasks: DevTask[]
  subtasks: Subtask[]
}

export interface PlanningRunResponse {
  agent_run: AgentRun
  artifact: Artifact
}

File: services/api/app/models/tasks.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel

from .agents import AgentRun
from .artifacts import Artifact
from .core import AssigneeType

DevTaskType = Literal[
    "backend", "frontend", "full_stack", "testing",
    "documentation", "infrastructure", "refactor", "unknown",
]
DevTaskStatus = Literal["proposed", "ready", "in_progress", "blocked", "completed"]
DevTaskPriority = Literal["low", "medium", "high"]
SubtaskStatus = Literal["proposed", "ready", "in_progress", "blocked", "completed"]


class DevTask(BaseModel):
    id: str
    project_id: str
    requirement_id: str | None = None
    ticket_id: str | None = None
    source_analysis_id: str | None = None
    agent_run_id: str
    epic_id: str | None = None
    title: str
    description: str
    task_type: DevTaskType = "unknown"
    status: DevTaskStatus = "proposed"
    priority: DevTaskPriority = "medium"
    sequence_order: int = 0
    depends_on: list[str] = []
    acceptance_criteria: list[str] = []
    definition_of_done: list[str] = []
    qa_required: bool = False
    suggested_agent_type: str | None = None
    assignee_type: AssigneeType = "unassigned"
    assignee_id: str | None = None
    assignee_name: str | None = None
    created_at: datetime
    updated_at: datetime


class Subtask(BaseModel):
    id: str
    dev_task_id: str
    project_id: str
    title: str
    description: str
    status: SubtaskStatus = "proposed"
    sequence_order: int = 0
    acceptance_criteria: list[str] = []
    qa_required: bool = False
    assignee_type: AssigneeType = "unassigned"
    assignee_id: str | None = None
    assignee_name: str | None = None
    created_at: datetime
    updated_at: datetime


class DevTaskUpdate(BaseModel):
    title: str | None = None
    description: str | None = None
    status: DevTaskStatus | None = None
    priority: DevTaskPriority | None = None
    sequence_order: int | None = None
    depends_on: list[str] | None = None
    acceptance_criteria: list[str] | None = None
    definition_of_done: list[str] | None = None
    qa_required: bool | None = None
    suggested_agent_type: str | None = None
    epic_id: str | None = None
    assignee_type: AssigneeType | None = None
    assignee_id: str | None = None
    assignee_name: str | None = None


class SubtaskUpdate(BaseModel):
    title: str | None = None
    description: str | None = None
    status: SubtaskStatus | None = None
    sequence_order: int | None = None
    acceptance_criteria: list[str] | None = None
    qa_required: bool | None = None
    assignee_type: AssigneeType | None = None
    assignee_id: str | None = None
    assignee_name: str | None = None


class DevTaskWithReadiness(DevTask):
    is_ready: bool = True
    blocked_by: list[str] = []


class TaskDecompositionRunCreate(BaseModel):
    provider: str | None = None
    expensive_approved: bool = False


class TaskDecompositionResponse(BaseModel):
    agent_run: AgentRun
    artifact: Artifact
    dev_tasks: list[DevTask]
    subtasks: list[Subtask]


class DevTaskWithSubtasksResponse(BaseModel):
    dev_task: DevTaskWithReadiness
    subtasks: list[Subtask]
